codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java

```diff
@@ -35,133 +32,79 @@
 import java.util.NoSuchElementException;
 import java.util.Set;
 import java.util.TimeZone;
-import java.util.TreeSet;
 
-import static io.netty.util.internal.ObjectUtil.*;
+import static io.netty.util.HashingStrategy.JAVA_HASHER;
+import static io.netty.util.internal.ObjectUtil.checkNotNull;
 
+/**
+ * Default implementation of {@link Headers};
+ *
+ * @param <T> the type of the header name and value.
+ */
 public class DefaultHeaders<T> implements Headers<T> {
     /**
-     * Allows users of this interface to specify a hash code other than the default {@link Object#hashCode()}
-     */
-    public interface HashCodeGenerator<T> {
-        /**
-         * Obtain the hash code for the given {@code name}
-         *
-         * @param name The name to generate the hash code for
-         * @return The hash code for {@code name}
-         */
-        int generateHashCode(T name);
-    }
-
-    /**
-     * Allows users to convert the {@code name} elements before being processed by this map
-     */
-    public interface NameConverter<T> {
-        /**
-         * Convert the {@code name} to some other form of the same object type
-         *
-         * @param name The object to convert
-         * @return The results of the conversion
-         */
-        T convertName(T name);
-    }
-
-    /**
-     * Uses the {@link #hashCode()} method to generate the hash code.
+     * How big the underlying array is for the hash data structure.
+     * <p>
+     * This should be a power of 2 so the {@link #index(int)} method can full address the memory.
      */
-    public static final class JavaHashCodeGenerator<T> implements HashCodeGenerator<T> {
-        @Override
-        public int generateHashCode(T name) {
-            return name.hashCode();
-        }
-    }
+    private static final int ARRAY_SIZE = 1 << 4;
```

Comment:
Yes I can experiment in the headers context, I will run the headers benchmark both ways.

The power of 2 was chosen because masking is faster than modulo (about 77% in my tests, just in case see jmh below).

```java
    @Param({ "16", "17", "31", "32" })
    private int size;

    private int value;

    @Setup(Level.Trial)
    public void setup() {
        value = 0;
    }

    @Benchmark
    public void modulo(Blackhole bh) {
        bh.consume((value++) % size);
    }

    @Benchmark
    public void mask(Blackhole bh) {
        bh.consume((value++) & size);
    }
```